.github/workflows/integration-test-containers.yml

```diff
@@ -0,0 +1,136 @@
+name: Build Integration Test Containers
+
+on:
+  workflow_call:
+    outputs:
+      qa-tag:
+        description: The QA tag used to build the containers
+        value: ${{ jobs.common-variables.outputs.qa-tag }}
+
+jobs:
+  test-containers:
+    runs-on: ubuntu-latest
+    outputs:
+      changed: ${{ steps.filter.outputs.container }}
+      changed_files: ${{ steps.filter.outputs.container_files }}
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: dorny/paths-filter@v2
+        id: filter
+        with:
+          list-files: shell
+
+          # Filter on changes strictly in subdirectories within
+          # integration-test/container
+          # (i.e. exclude files in integration-tests/container itself)
+          filters: |
+            container:
+              - integration-tests/container/*/**
+
+  common-variables:
+    needs: test-containers
+    runs-on: ubuntu-latest
+    outputs:
+      qa-tag: ${{ steps.variables.outputs.qa-tag }}
+      build-dirs: ${{ steps.variables.outputs.build-dirs}}
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Set variables
+        id: variables
+        run: |
+          if [[ "${{ contains(github.event.pull_request.labels.*.name, 'build-qa-containers') }}" == 'true' ]]; then
+            #
+            # If we're building all qa images (on a PR) we have to build everything, because the new images
+            # are tagged, and that tag is passed to the integration test and applied to all images.
+            #
+            dirs=$(find integration-tests/container/ -type d -mindepth 1 -maxdepth 1)
+            qa_tag="$(echo "${GITHUB_SHA}" | cut -c1-8)"
+          else
+            #
+            # On master, we have the luxury of only building things that have changed,
+            # because the QA tag is not applied (instead we'll update the floating tags)
+            #
+            dirs=$(dirname ${{ needs.test-containers.outputs.changed_files }} | sort | uniq)
```

Comment:
We will have to remember that subdirectories are forbidden here, this is probably acceptable.